### PR TITLE
Introduce Microsoft.App/builders tracked resource and Microsoft.App/builds proxy resource

### PR DESCRIPTION
**Pull request by pauld-msft**
## ARM API Information (Control Plane)
------------
These new resource types were already introduced in these pull requests: https://github.com/Azure/azure-rest-api-specs/pull/23697 (merged) and https://github.com/Azure/azure-rest-api-specs/pull/24050 (abandoned in favor of this).

However, they were created under the api version `2023-05-01-preview`, and we have since decided to change the version to `2023-05-02-preview`. This PR is pointed at the release branch for `2023-05-02-preview`, which is based off of the current `2023-04-01-preview` api version.

------------

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[x] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version). Note that this doesn't apply if you are trying to merge a PR that was previously in the private repository.

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/overview_breakingchanges).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](http://aka.ms/aprwiki))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (mentat9) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json:45]: >v3 [](http://example.com/codeflow?start=69&length=2)

Recommend tracking to latest version of common-types (currently v5).

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (mentat9) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json:61]: >./CommonDefinitions.json#/definitions/DefaultErrorResponse [](http://example.com/codeflow?start=23&length=58)

Please consider migrating to the ErrorResponse definition in common-types. Not a requirement for this PR, but something you should consider for the future.

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (mentat9) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json:135]: >builderName [](http://example.com/codeflow?start=21&length=11)

Consider factoring this parameter out to the parameters section for use/consistency across multiple APIs.

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (mentat9) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json:477]: >InProgress [](http://example.com/codeflow?start=9&length=10)

When is this state used? Consider something more specific (e.g. Creating, Initializing, Updating, or similar).

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builds.json
Comment (mentat9) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builds.json:116]: >buildName [](http://example.com/codeflow?start=21&length=9)

Consider moving this (and builderName) to parameters {} section for use/consistency across multiple APIs.

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builds.json
Comment (mentat9) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builds.json:430]: >identity [](http://example.com/codeflow?start=9&length=8)

Don't reuse names of top-level properties within the property bag. Maybe identityResourceId, containerAccessIdentityResourceId or similar.

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (pauld-msft) [reply specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json:61]: For these resource types, we are trying to maintain consistency with the rest of the Microsoft.App resource types, so would prefer to keep as is for now until the update is made across the RP

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (pauld-msft) [reply specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json]: updated to be creating

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builds.json
Comment (pauld-msft) [reply specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builds.json]: updated to identityResourceId

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (mentat9) [reply specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json]: Ideally, you could distinguish Creating and Updating and have both since PUT can be either a create or update. Just Creating is also OK, not blocking ARM signoff.

**Comment by mentat9** [comment]
@pauld-msft - Signed off for ARM with comments. Please consider one additional suggestion.

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (pauld-msft) [reply specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json]: That makes sense. I'll add `updating` as well. Just want to clarify one thing quickly.

The [RPC ](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#provisioningstate-property) mentions that there are the three terminal states. It sounds like we can have a scenario (like when we are `updating`), when the resource is able to transition out of the terminal state (of `succeeded`) and into another non-terminal state (`updating`), and then eventually back into the terminal state. 

Is this correct? I was originally interpreting a terminal state as a "final" state from which the resource would not be able to transition, but sounds like that is not right. Is it just terminal for each async operation, and not the resource as a whole?

Comment (mentat9) [reply specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json]: Correct. I think terminal here means it won't be changed by the RP, only in response to a client request.

**Comment by mentat9** [comment]
> @mentat9 - Thanks for the signoff. I did just make one minor update to the API, hope its not an issue. Basically, instead of having customers pass in managed identity to use for a given container registry every time they create a build object, we now have a dictionary in the builder (parent) object that will track the mapping, as it theoretically should not change very much.

@pauld-msft - I took a look and it's fine: no need to reset ARM signoff state.

**Comment by pauld-msft** [comment]
@kazrael2119 @mentat9 I believe this is ready to be merged into our feature branch if it looks good to you all

**Comment by kazrael2119** [comment]
I don't know how to deal with ```SuppressionReviewRequired``` label, @raych1 ,can you give some advice? thanks

File: specification/app/resource-manager/readme.md
Comment (kazrael2119) [inline specification/app/resource-manager/readme.md:59]: @mentat9 can you help review this? beacause "SuppressionReviewRequired" label is added, thanks

Comment (pauld-msft) [reply specification/app/resource-manager/readme.md:59]: Links to reasoning for the suppressions: 

- https://github.com/Azure/azure-rest-api-specs/pull/23697#issuecomment-1545915040
- https://github.com/Azure/azure-rest-api-specs-pr/pull/10719#issuecomment-1385705675

Comment (kazrael2119) [reply specification/app/resource-manager/readme.md:59]: @pauld-msft , If you want to merge this pr asap, I think you can ask ArmReview Board to review this suppression. After "Approved Suppression" label is added , I think this pr is ok to merge

**Comment by mentat9** [comment]
Tagging @ms-zhenhua who is oncall this week.

File: specification/app/resource-manager/readme.md
Comment (ms-zhenhua) [reply specification/app/resource-manager/readme.md:59]: @pauld-msft @kazrael2119  approved and signed off

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (kazrael2119) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json:497]: can you add ```"x-ms-parameter-location": "method"``` in ```BuilderNameParameter``` and ```BuildNameParameter```

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builds.json
Comment (kazrael2119) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builds.json:482]: can you add ```"x-ms-parameter-location": "method"``` in ```BuildNameParameter```

File: specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json
Comment (kazrael2119) [inline specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Builders.json:45]: I think it should be v3 not v5 to keep same with other services:https://github.com/Azure/azure-rest-api-specs/blob/release-Microsoft.App-2023-05-02-preview/specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/ContainerApps.json#L27

**Comment by pauld-msft** [comment]
@kazrael2119 and @ms-zhenhua, I would appreciate a re-review, and then merge in if all looks good. Thanks!

**Review by kazrael2119 — APPROVED** [review]
skip sdk breaking change because it's not to main. I want to review sdk breaking change when release branch need to merge to main

**Comment by kazrael2119** [comment]
> @kazrael2119 - ok, looks like we have the arm signoff and suppression approved

@pauld-msft ,please let me know when you want to merge this pr, thanks

**Pull request merged**
